### ISSUE TITLE
Archives of a commit have different md5sum each time even though content stays the same

### ISSUE THREAD

**Issue by hv15** [gitbucket/gitbucket#1333]
## Issue

**Impacted version**: 4.5 and 4.2.1 (upgrade to see if problem persisted)

**Deployment mode**: standalone app

**Problem description**:
Whenever I download an archive (tar.gz and/or zip) at a particular commit, the resulting file
has a different checksum (md5sum) every time, even though there is no apparent difference
in the content of the file. _This problem is probably caused by gzip (or zip) adding a new time stamp within the archive header region, resulting in a different checksum_.

**For Example**:
I have a repo [startw](https://git.hans.ninja/hans/startw) which has several tags - if I download an
archive of the latest tag (e.g. using the link https://git.hans.ninja/hans/startw/archive/1.1.1.tar.gz) with
curl and do an md5sum, I get the following:

``` sh
$ cd /tmp
$ curl -OL https://git.hans.ninja/hans/startw/archive/1.1.1.tar.gz
$ md5sum 1.1.1.tar.gz
8a5a94d54cf8b5740cab8376102a1edc  1.1.1.tar.gz
```

If I repeat the above procedure, I get a different checksum result:

``` sh
$ cd /tmp
$ curl -OL https://git.hans.ninja/hans/startw/archive/1.1.1.tar.gz
$ md5sum 1.1.1.tar.gz
e2dfae2e18465f80b0ae73435483ceef  1.1.1.tar.gz
```

Obviously there must be something different about the archives to cause this, and so I decided
to do a md5sum of the contents of the archives I downloaded - for clarity lets call them _test1_ and _test2_ respectively.

``` sh
$ md5sum 1.1.1.tar.gz 
e2dfae2e18465f80b0ae73435483ceef  1.1.1.tar.gz

$ tar xvf 1.1.1.tar.gz -C test1
LICENSE
README
example.conf/
example.conf/winitrc
man/
man/startw.1
startw

$ curl -OL https://git.hans.ninja/hans/startw/archive/1.1.1.tar.gz
$ md5sum 1.1.1.tar.gz                                             
3d1f874b98c20d605438c981b6372aff  1.1.1.tar.gz

$ tar xvf 1.1.1.tar.gz -C test2                                   
LICENSE
README
example.conf/
example.conf/winitrc
man/
man/startw.1
startw
```

If we now have a look, we can see that the content of both archives is identical:

``` sh
$ cd test1
$ find . -type f -exec md5sum {} \;
c5851fb713b928ac2c99b6f513361762  ./startw
9ef65ab9c6f9aa9b2e06ab410117264f  ./man/startw.1
1990549d86e2944bc8569dbb2ef1ab6b  ./example.conf/winitrc
ed940bce4db3b092b77bdf752a10f991  ./README
704901c775dac923198527e631cf7d73  ./LICENSE

$ cd test2
$ find . -type f -exec md5sum {} \;
c5851fb713b928ac2c99b6f513361762  ./startw
9ef65ab9c6f9aa9b2e06ab410117264f  ./man/startw.1
1990549d86e2944bc8569dbb2ef1ab6b  ./example.conf/winitrc
ed940bce4db3b092b77bdf752a10f991  ./README
704901c775dac923198527e631cf7d73  ./LICENSE
```

This implies that something is happening with the actual generation of the tar.gz archive files themselves. I presume it has to do with gzip and it setting a modification time stamp within the header.
Having a look at the headers, this seems to be the case:

``` sh
$ gzip -lv 1.1.1.tar.gz 
method  crc     date  time           compressed        uncompressed  ratio uncompressed_name
defla 78e3b08b Oct 24 14:00                3360               20480  83.7% 1.1.1.tar

$ curl -OL https://git.hans.ninja/hans/startw/archive/1.1.1.tar.gz
$ gzip -lv 1.1.1.tar.gz                                           
method  crc     date  time           compressed        uncompressed  ratio uncompressed_name
defla 62e327b0 Oct 24 14:02                3359               20480  83.7% 1.1.1.tar
```

The problem is the same with ZIP archives as well.

**Comment by ritschwumm** [comment]
why is this a bug?

**Comment by McFoggy** [comment]
I put the bug label ont this issue because by reading the issue I quickly thought that a materialization of something that does not change should not change  ; but you're right nothing enforces that.

**Comment by McFoggy** [comment]
@hv15 would your requirement be covered by a release page like described in #607

**Comment by ritschwumm** [comment]
@hv15 let's assume the inside date stamp would be constant between builds - which date should that be?

**Comment by hv15** [comment]
@McFoggy, that would help, but in the case where I want to download a tar.gz of a particular commit, the problem would still persist. My context is automated distro package compiling/building whereby a checksum (or signature) is the primary way of confirming that the files are the files I need/want to build.

@ritschwumm I don't understand the question, perhaps due to my ignorance on how you actually generate the archive. As I understand it one needs to specify that no timestamp is included within the archive header - that would resolve the changing checksum problem I think.

**Comment by McFoggy** [comment]
@hv15 I guess that if gitbucket had a release page in which you could publish artifacts then it would solve your problem. You would be able then to publish a zip/jar/war or whatever and its associated md5 or sha1.
Being able 'just' to download the repository content as a zip or tar.gz does not cover a lot of release scenario.

@ritschwumm if we wanted to, we could pick up the date of the commit for example.

**Comment by ritschwumm** [comment]
hm.. had a short look at the tgz generation, and it doesn't seem you can remove that timestamp. you could set it to 0 (unix epoch) by some hackery in TgzFormat.

something like this - but Tgz is final so it wouldn't compile.
```scala
ArchiveCommand.registerFormat("tar.gz", new TgzFormat {
    override def createArchiveOutputStream(s:OutputStream):ArchiveOutputStream  = {
      val params = new GzipParameters
      params setModificationTime 0L
      val out = new GzipCompressorOutputStream(s)
      tarFormat.createArchiveOutputStream(out)
    }
  })
```

**Comment by hv15** [comment]
@McFoggy I see your point, but it is a release scenario nonetheless. Your idea on using the commit timestamp I think it a good solution to this.

**Comment by McFoggy** [comment]
> but it is a release scenario nonetheless
@hv15 a release scenario of what? The functionality your are using is just a download feature of some source code extracted from a git repository on a given commit. In no way it is a release feature.

But It can be an enhancement to say that the archive exposed via the download functionalities take the modification date of the commit they represent  ; I would assume this can be a good compromise.

**Comment by hv15** [comment]
I've had a look at the zip archive solution, and similar to @ritschwumm solution for tar.gz, one would have to call `ZipArchiveEntry.setTime()` to set the last modified timestamp for the archive.

**Comment by takezoe** [comment]
This is a JGit issue. I sent pull request to the JGit project: https://github.com/eclipse/jgit/pull/41

**Comment by takezoe** [comment]
The JGit project doesn't accept pull requests on GitHub, so I pushed to the Gerrit code review: https://git.eclipse.org/r/#/c/91116/

**Comment by takezoe** [comment]
Fixed in JGit 4.7.0.
https://projects.eclipse.org/projects/technology.jgit/releases/4.7.0
